core/src/main/java/com/linecorp/armeria/server/ExceptionHandler.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static com.linecorp.armeria.server.ExceptionHandlerUtil.internalServerErrorResponse;
+import static com.linecorp.armeria.server.ExceptionHandlerUtil.serviceUnavailableResponse;
+
+import java.util.function.BiFunction;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+
+/**
+ * Converts a {@link Throwable} to an {@link AggregatedHttpResponse}.
+ * Use this {@link ExceptionHandler} to send a different response depending on the {@link Throwable}:
+ *
+ * <pre>{@code
+ * ExceptionHandler handler = (ctx, cause) -> {
+ *     if (cause instanceof IllegalArgumentException) {
+ *         return AggregatedHttpResponse.of(HttpStatus.BAD_REQUEST);
+ *     }
+ *
+ *     // You can return a different response using the path.
+ *     if ("/foo".equals(ctx.path())) {
+ *         return AggregatedHttpResponse.of(HttpStatus.OK, MediaType.PLAIN_TEXT, "Foo exception");
+ *     }
+ *
+ *     // Delegate to the default handler.
+ *     return ExceptionHandler.ofDefault().apply(ctx, cause);
+ * }
+ *
+ * Server.builder().exceptionHandler(handler)...
+ *
+ * }</pre>
+ *
+ * @see ServerBuilder#exceptionHandler(ExceptionHandler)
+ */
+@FunctionalInterface
+public interface ExceptionHandler extends BiFunction<ServiceRequestContext, Throwable, AggregatedHttpResponse> {
```

Comment:
Is it better to allow to return a null value and apply the default exception handler if a null value is returned?

Reply:
Yeah, that was one of the hardest choices to make.
So I just looked through other interfaces and it seems like that returning `null` for these sorts of interfaces is not conventional. 🤔

Reply:
- Can we return an `HttpResponse` instead of an aggregated version? e.g. a user might want to send out a large response although it's not very common.
- If we allow returning `null`, we could add methods like `orElse()` :smile: Sounds good to me.

Reply:
>  So I just looked through other interfaces and it seems like that returning null for these sorts of interfaces is not conventional. 🤔

`GrpcStatusFunction` has already allowed returning a null value.
https://github.com/line/armeria/blob/480321970111ac6249775199fedadcc71b862e84/grpc/src/main/java/com/linecorp/armeria/common/grpc/GrpcStatusFunction.java#L33-L43

Reply:
> GrpcStatusFunction has already allowed returning a null value. 😀

I missed that. 😅 
Yeah, let me return an `HttpResponse` or `null` so that we can chain the functions. Thanks!

Reply:
Had a chat with @trustin and @ikhoon, and we decided to keep returning `AggregatedHttpRespons` because:
- we should send the response immediately when a request timed out but if we use `HttpResponse` we might not able to send the response right away.
- we need another `HttpResponseSubscriber` in the `HttpResponseSubscriber` which makes it complicated.

We might revisit this later if there's some demand to return `HttpResposne`.